storage: transfer away leases when draining

asubiotto opened the pull request:
When a Store is put into draining mode, it now attempts to transfer away
any leases owned by its Replicas. This avoids the need to wait for
leases to expire while maintaining availability.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13792)
<!-- Reviewable:end -->

Commit:
storage: transfer away leases when draining

When a Store is put into draining mode, it now attempts to transfer away
any leases owned by its Replicas. This avoids the need to wait for
leases to expire while maintaining availability.

asubiotto commented:
Currently working on some code reuse in `pkg/storage/client_replica_test.go` so don't pay the test files too much attention yet.

tamird commented:
Reviewed 4 of 6 files at r1.
Review status: 4 of 6 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/storage/store.go, line 969 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-Kdlur75TIMoNiQLHJJv:-Kdlur75TIMoNiQLHJJw:b-s26hvn) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ceda78743f0a917c4ead3d53556391fa973e07f/pkg/storage/store.go#L969)):*
> ```Go
> }
> 
> // SetDraining (when called with 'true') prevents all of the Store's
> ```

"SetDraining causes incoming lease transfers to be rejected and prevents ..." instead of the second sentence, which is confusing since it's not clear who is attempting the transfer.

---

*[pkg/storage/store.go, line 991 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-KdlvjbJ7SNuyn2YfcxA:-KdlvjbJ7SNuyn2YfcxB:bpw0xua) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ceda78743f0a917c4ead3d53556391fa973e07f/pkg/storage/store.go#L991)):*
> ```Go
> 			// If a lease request is in progress for us, we wait on it.
> 			if nextLease != nil && nextLease.OwnedBy(s.StoreID()) {
> 				r.mu.Lock()
> ```

this doesn't seem ok - i think you need all this jazz to be done under the same lock as the stuff going on in `getLease`. At least, that's what all existing code looks like.

---

*[pkg/storage/store.go, line 1002 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-Kdlw3QynRUbn32lkA4E:-Kdlw3QynRUbn32lkA4F:b5rjfli) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ceda78743f0a917c4ead3d53556391fa973e07f/pkg/storage/store.go#L1002)):*
> ```Go
> 
> 		if drainingLease.OwnedBy(s.StoreID()) && r.IsLeaseValid(drainingLease, s.Clock().Now()) {
> 			var zone config.ZoneConfig
> ```

`zone := config.DefaultZoneConfig()` and then overwrite it below if necessary?

---

*[pkg/storage/store.go, line 1008 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-KdlvyRWjFOuMLzAABaa:-KdlvyRWjFOuMLzAABab:bke1nhw) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ceda78743f0a917c4ead3d53556391fa973e07f/pkg/storage/store.go#L1008)):*
> ```Go
> 				zone, err = sysCfg.GetZoneConfigForKey(desc.StartKey)
> 				if log.V(1) && err != nil {
> 					log.Errorf(context.TODO(), "could not get zone config for key %s when draining: %s", desc.StartKey, err)
> ```

extract a single context and annotate it.

also, this needs to stop executing, i think.

---

*[pkg/storage/store.go, line 1016 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-KdlvukhqT9lwVd6ADUO:-KdlvukhqT9lwVd6ADUP:b-siacsw) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ceda78743f0a917c4ead3d53556391fa973e07f/pkg/storage/store.go#L1016)):*
> ```Go
> 				context.TODO(),
> 				r,
> 				r.Desc(),
> ```

why doesn't this use `desc`?

---

*[pkg/storage/store.go, line 1021 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-KdlwAnHf5ONcqP9iBLR:-KdlwAnHf5ONcqP9iBLS:bk5udjr) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ceda78743f0a917c4ead3d53556391fa973e07f/pkg/storage/store.go#L1021)):*
> ```Go
> 				false, /* checkCandidateFullness */
> 			); log.V(1) && err != nil {
> 				log.Errorf(context.TODO(), "error transferring lease when draining: %s", err)
> ```

also needs to stop execution?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 5 of 6 files at r1.
Review status: 5 of 6 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/storage/store.go, line 975 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-KdxQD62S-tKm7bl_aer:-KdxQD63u2x1iey8GUGp:bdn546v) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ceda78743f0a917c4ead3d53556391fa973e07f/pkg/storage/store.go#L975)):*
> ```Go
> // When called with 'false', returns to the normal mode of operation.
> //
> // TODO(asubiotto): Return an error when there may be active leases held by some
> ```

What is the goal of this TODO? What will we do with this error if we don't abort the draining process? (in general aborting the draining process seems like a bad idea - in most cases you'd rather exit gracelessly than have the process stay alive when you're trying to take it down)

---

*[pkg/storage/store.go, line 985 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-KdxTOjL1sRUMR2eQ2Lp:-KdxTOjL1sRUMR2eQ2Lq:b-iav2x5) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ceda78743f0a917c4ead3d53556391fa973e07f/pkg/storage/store.go#L985)):*
> ```Go
> 
> 	sysCfg, sysCfgSet := s.cfg.Gossip.GetSystemConfig()
> 	newStoreReplicaVisitor(s).Visit(func(r *Replica) bool {
> ```

Transferring all leases serially will take a long time; we need to be able to have multiple transfers in flight at once. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Addressed comments and moved the tests into `TestRangeTransferLease`. PTAL.

---

Review status: 5 of 6 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/storage/store.go, line 969 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-Kdlur75TIMoNiQLHJJv:-Ke4qa21kX9drv6IXs3W:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ceda78743f0a917c4ead3d53556391fa973e07f/pkg/storage/store.go#L969)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

"SetDraining causes incoming lease transfers to be rejected and prevents ..." instead of the second sentence, which is confusing since it's not clear who is attempting the transfer.
</blockquote></details>

Done.

---

*[pkg/storage/store.go, line 975 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-KdxQD62S-tKm7bl_aer:-Ke4qeVNYzg4q1lu0hhX:bjuiiqm) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ceda78743f0a917c4ead3d53556391fa973e07f/pkg/storage/store.go#L975)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

What is the goal of this TODO? What will we do with this error if we don't abort the draining process? (in general aborting the draining process seems like a bad idea - in most cases you'd rather exit gracelessly than have the process stay alive when you're trying to take it down)
</blockquote></details>

The goal is to provide the caller with more information as to the state that SetDraining left the store in. I agree, the draining process shouldn't be aborted on error, which is why I suggest changing the upper-level draining logic. I don't have strong feelings about this though and it might be better to not return errors since this is a best-effort attempt.

---

*[pkg/storage/store.go, line 985 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-KdxTOjL1sRUMR2eQ2Lp:-Ke53wVXL-6wlIeXteaJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ceda78743f0a917c4ead3d53556391fa973e07f/pkg/storage/store.go#L985)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Transferring all leases serially will take a long time; we need to be able to have multiple transfers in flight at once. 
</blockquote></details>

Done.

---

*[pkg/storage/store.go, line 991 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-KdlvjbJ7SNuyn2YfcxA:-Ke5TOYtl9OHmB59gyAT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ceda78743f0a917c4ead3d53556391fa973e07f/pkg/storage/store.go#L991)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this doesn't seem ok - i think you need all this jazz to be done under the same lock as the stuff going on in `getLease`. At least, that's what all existing code looks like.
</blockquote></details>

Done.

---

*[pkg/storage/store.go, line 1002 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-Kdlw3QynRUbn32lkA4E:-Ke54kwq4P0y_wLGM3XP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ceda78743f0a917c4ead3d53556391fa973e07f/pkg/storage/store.go#L1002)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

`zone := config.DefaultZoneConfig()` and then overwrite it below if necessary?
</blockquote></details>

Done.

---

*[pkg/storage/store.go, line 1008 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-KdlvyRWjFOuMLzAABaa:-Ke57D8KxihDIIH1aNVy:b-l23yd6) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ceda78743f0a917c4ead3d53556391fa973e07f/pkg/storage/store.go#L1008)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

extract a single context and annotate it.

also, this needs to stop executing, i think.
</blockquote></details>

Why do you think it needs to stop executing? I think that if a zone config does not exist, it's better to transfer the lease with no constraints than not to. 

---

*[pkg/storage/store.go, line 1016 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-KdlvukhqT9lwVd6ADUO:-Ke575cNaFsbBnUmwMjH:b-v0flhd) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ceda78743f0a917c4ead3d53556391fa973e07f/pkg/storage/store.go#L1016)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why doesn't this use `desc`?
</blockquote></details>

Done. Was a mistake.

---

*[pkg/storage/store.go, line 1021 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-KdlwAnHf5ONcqP9iBLR:-Ke58lVDD1vMErkuqfk_:b-2x5jp6) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ceda78743f0a917c4ead3d53556391fa973e07f/pkg/storage/store.go#L1021)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

also needs to stop execution?
</blockquote></details>

If a replica fails to transfer away a lease, I don't think we should stop other replicas from attempting to transfer away their leases.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-:-Ke5ZKQZG6mnCzZMGu7w:b-far0k1)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 4 files at r2.
Review status: 6 of 7 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/storage/store.go, line 993 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-Ke5q1nJeenvZ6Y0xIb4:-Ke5q1nJeenvZ6Y0xIb5:b-yo2aaa) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc2bd405e55d05e1d8d1b8ff002e6a85c56ea431/pkg/storage/store.go#L993)):*
> ```Go
> 			var drainingLease *roachpb.Lease
> 			for {
> 				if requestInProgress := func() <-chan *roachpb.Error {
> ```

why is this function necessary? this defer doesn't seem to buy much:
```
var leaseCh <-chan *roachpb.Error
r.mu.Lock()
lease, nextLease := r.getLeaseRLocked()
if nextLease != nil && nextLease.OwnedBy(s.StoreID()) {
  leaseCh = r.mu.pendingLeaseRequest.JoinRequest()
}
r.mu.Unlock()

if leaseCh != nil {
  <-leaseCh
  continue
}
drainingLease = lease
break
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 4 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/storage/client_replica_test.go, line 531 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-Ke5t7Y0k5kXBQdpgbsH:-Ke5t7Y0k5kXBQdpgbsI:b-fejbjx) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc2bd405e55d05e1d8d1b8ff002e6a85c56ea431/pkg/storage/client_replica_test.go#L531)):*
> ```Go
> 	checkHasLease := func(t *testing.T, sender *storage.Stores) {
> 		testutils.SucceedsSoon(t, func() error {
> 			if pErr := sendRead(sender); pErr != nil {
> ```

`return sendRead(sender).GoError()` - this does the right thing for nil.

---

*[pkg/storage/client_replica_test.go, line 574 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-Ke5tTorxltU7Lc1uRml:-Ke5tTosqSnONPWP189N:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc2bd405e55d05e1d8d1b8ff002e6a85c56ea431/pkg/storage/client_replica_test.go#L574)):*
> ```Go
> 		shouldRenewTS := lease.Expiration.Add(-1, 0)
> 		mtc.manualClock.Set(shouldRenewTS.WallTime + 1)
> 		if pErr := sendRead(sender); pErr != nil {
> ```

ditto

---

*[pkg/storage/client_replica_test.go, line 685 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-Ke5u0cQ6eaI9DZETcFE:-Ke5u0cQ6eaI9DZETcFF:b-rzu1ga) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc2bd405e55d05e1d8d1b8ff002e6a85c56ea431/pkg/storage/client_replica_test.go#L685)):*
> ```Go
> 		desc, err := r.GetReplicaDescriptor()
> 		if err != nil {
> 			return errors.New("could not get replica descriptor")
> ```

why not use the error returned?

---

*[pkg/storage/client_replica_test.go, line 688 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-Ke5u36Y96bwZtn_xWf_:-Ke5u36Y96bwZtn_xWfa:b-edxlaz) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc2bd405e55d05e1d8d1b8ff002e6a85c56ea431/pkg/storage/client_replica_test.go#L688)):*
> ```Go
> 		}
> 		if desc != replica0Desc {
> 			return errors.New("replica0 not raft leader")
> ```

please consider what someone would do with this message. which replica is the raft leader? in general, error messages should be useful, and this one is not.

---

*[pkg/storage/client_replica_test.go, line 697 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-Ke5uEEhhlfWif_9zZiP:-Ke5uEEhhlfWif_9zZiQ:br74agx) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc2bd405e55d05e1d8d1b8ff002e6a85c56ea431/pkg/storage/client_replica_test.go#L697)):*
> ```Go
> 		progress, ok := status.Progress[uint64(replica1Desc.ReplicaID)]
> 		if !ok {
> 			return errors.New("could not get replica1 progress")
> ```

ditto. include the map?

---

*[pkg/storage/client_replica_test.go, line 700 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-Ke5uGRHhAcnbxbL4V_9:-Ke5uGRHhAcnbxbL4V_A:b84pm5n) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc2bd405e55d05e1d8d1b8ff002e6a85c56ea431/pkg/storage/client_replica_test.go#L700)):*
> ```Go
> 		}
> 		if progress.Match < status.Commit {
> 			return errors.New("replica1 failed to catch up")
> ```

ditto, include something relevant?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/storage/client_replica_test.go, line 697 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-Ke5uEEhhlfWif_9zZiP:-Ke6yMaFlT7ddV0PsWU-:b-bibmvv) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc2bd405e55d05e1d8d1b8ff002e6a85c56ea431/pkg/storage/client_replica_test.go#L697)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

ditto. include the map?
</blockquote></details>

What map? It's nil/undefined if `!ok`. There's not much useful we could add here.

---

*[pkg/storage/client_replica_test.go, line 700 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-Ke5uGRHhAcnbxbL4V_9:-Ke6yXmNV-u27SqKK-G9:b-2wiix9) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc2bd405e55d05e1d8d1b8ff002e6a85c56ea431/pkg/storage/client_replica_test.go#L700)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

ditto, include something relevant?
</blockquote></details>

This seems fine to me; the exact values are unlikely to be helpful.

---

*[pkg/storage/store.go, line 975 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-KdxQD62S-tKm7bl_aer:-Ke6vp0Ro-JEkFPJ065y:baqlvy2) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ceda78743f0a917c4ead3d53556391fa973e07f/pkg/storage/store.go#L975)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

The goal is to provide the caller with more information as to the state that SetDraining left the store in. I agree, the draining process shouldn't be aborted on error, which is why I suggest changing the upper-level draining logic. I don't have strong feelings about this though and it might be better to not return errors since this is a best-effort attempt.
</blockquote></details>

It's probably better to do that in some other way than returning an error. Look at the way the FreezeCluster stuff works, by returning status information as the freeze progresses.

---

*[pkg/storage/store.go, line 985 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-KdxTOjL1sRUMR2eQ2Lp:-Ke6w7bd3UGBdOKDSiem:b-7rmci) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ceda78743f0a917c4ead3d53556391fa973e07f/pkg/storage/store.go#L985)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Transferring every lease at once is probably also a bad idea. Use Stopper.RunLimitedAsyncTask instead.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-:-Ke6z2eLqWmEv8D_u7Rd:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/storage/client_replica_test.go, line 697 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-Ke5uEEhhlfWif_9zZiP:-KeOMJc3tDrBmZgcANZQ:b96isbe) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc2bd405e55d05e1d8d1b8ff002e6a85c56ea431/pkg/storage/client_replica_test.go#L697)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

What map? It's nil/undefined if `!ok`. There's not much useful we could add here.
</blockquote></details>

"the map" I was referring to is `status.Progress`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/storage/client_replica_test.go, line 531 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-Ke5t7Y0k5kXBQdpgbsH:-KeJbimrROqvG3mJud65:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc2bd405e55d05e1d8d1b8ff002e6a85c56ea431/pkg/storage/client_replica_test.go#L531)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

`return sendRead(sender).GoError()` - this does the right thing for nil.
</blockquote></details>

Done.

---

*[pkg/storage/client_replica_test.go, line 574 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-Ke5tTorxltU7Lc1uRml:-KeJbq_VZfxpX5KE5EvF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc2bd405e55d05e1d8d1b8ff002e6a85c56ea431/pkg/storage/client_replica_test.go#L574)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---

*[pkg/storage/client_replica_test.go, line 685 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-Ke5u0cQ6eaI9DZETcFE:-KeJkZyx3yN4IQ52BKjO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc2bd405e55d05e1d8d1b8ff002e6a85c56ea431/pkg/storage/client_replica_test.go#L685)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why not use the error returned?
</blockquote></details>

Done.

---

*[pkg/storage/client_replica_test.go, line 688 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-Ke5u36Y96bwZtn_xWf_:-KeJi2_yLp-hnIpHKQFO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc2bd405e55d05e1d8d1b8ff002e6a85c56ea431/pkg/storage/client_replica_test.go#L688)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

please consider what someone would do with this message. which replica is the raft leader? in general, error messages should be useful, and this one is not.
</blockquote></details>

Done.

---

*[pkg/storage/client_replica_test.go, line 697 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-Ke5uEEhhlfWif_9zZiP:-KeZCVijMIfeA1B1LpkL:bflprfq) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc2bd405e55d05e1d8d1b8ff002e6a85c56ea431/pkg/storage/client_replica_test.go#L697)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

"the map" I was referring to is `status.Progress`
</blockquote></details>

Added the map.

---

*[pkg/storage/store.go, line 975 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-KdxQD62S-tKm7bl_aer:-KeJmF94O84lAnQKztkN:b2m321j) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ceda78743f0a917c4ead3d53556391fa973e07f/pkg/storage/store.go#L975)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

It's probably better to do that in some other way than returning an error. Look at the way the FreezeCluster stuff works, by returning status information as the freeze progresses.
</blockquote></details>

Thanks for that, that will probably be a good thing to add to the whole draining process. Removed the error and TODO.

---

*[pkg/storage/store.go, line 985 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-KdxTOjL1sRUMR2eQ2Lp:-KeJwqfgddM9kOt6iyA0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ceda78743f0a917c4ead3d53556391fa973e07f/pkg/storage/store.go#L985)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Transferring every lease at once is probably also a bad idea. Use Stopper.RunLimitedAsyncTask instead.
</blockquote></details>

Done.

---

*[pkg/storage/store.go, line 993 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-Ke5q1nJeenvZ6Y0xIb4:-KeJlGoVcpSDMGcuavc9:boyexbg) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc2bd405e55d05e1d8d1b8ff002e6a85c56ea431/pkg/storage/store.go#L993)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why is this function necessary? this defer doesn't seem to buy much:
```
var leaseCh <-chan *roachpb.Error
r.mu.Lock()
lease, nextLease := r.getLeaseRLocked()
if nextLease != nil && nextLease.OwnedBy(s.StoreID()) {
  leaseCh = r.mu.pendingLeaseRequest.JoinRequest()
}
r.mu.Unlock()

if leaseCh != nil {
  <-leaseCh
  continue
}
drainingLease = lease
break
```
</blockquote></details>

Done. Thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/storage/client_raft_test.go, line 1035 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-KeZcdlHOPiA5KOJIYy7:-KeZcdlI2gt_KrnBVTyK:bkvp6y8) ([raw file](https://github.com/cockroachdb/cockroach/blob/5930241d3ce32a3caaeaf0b0405049f4c26c5fdc/pkg/storage/client_raft_test.go#L1035)):*
> ```Go
> 				go func(i int) {
> 					mtc.stores[i].SetDraining(true)
> 					wg.Done(nil)
> ```

seems that you can make this a plain `WaitGroup` now rather than a `WaitGroupWithErr`

---

*[pkg/storage/client_replica_test.go, line 685 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-Ke5u0cQ6eaI9DZETcFE:-KeZcCXfyj7ARNOpPjBS:b-yjfk5n) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc2bd405e55d05e1d8d1b8ff002e6a85c56ea431/pkg/storage/client_replica_test.go#L685)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

use `errors.Wrap` instead, or just `err`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Fixed a failing interactive tcl test.

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/storage/client_raft_test.go, line 1035 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-KeZcdlHOPiA5KOJIYy7:-Ke_GDNnhfvhVTQiVS5v:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5930241d3ce32a3caaeaf0b0405049f4c26c5fdc/pkg/storage/client_raft_test.go#L1035)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

seems that you can make this a plain `WaitGroup` now rather than a `WaitGroupWithErr`
</blockquote></details>

Done.

---

*[pkg/storage/client_replica_test.go, line 685 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-Ke5u0cQ6eaI9DZETcFE:-Ke_F3bEpMBaAHqRYK0S:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc2bd405e55d05e1d8d1b8ff002e6a85c56ea431/pkg/storage/client_replica_test.go#L685)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

use `errors.Wrap` instead, or just `err`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-:-KedMNg2IAXRA0UkMlIM:b-746wc1)*
<!-- Sent from Reviewable.io -->

tamird commented:
Maybe @knz wants to look

---

Reviewed 8 of 8 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-:-KedSClJBnhkgCJ5CfD4:b-vkjaw5)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 1 of 8 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/cli/interactive_tests/test_server_sig.tcl, line 48 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792#-KeeJK9qqhJpOrhDTJj9:-KeeJK9rX_2BplSMFaFt:bb6mt9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2d9bdbde48b35ad10fd145a0f39eb462cecb0d/pkg/cli/interactive_tests/test_server_sig.tcl#L48)):*
> ```Tcl
>     "shutdown completed" {}
>     timeout {exit 1}
> }
> ```

:+1:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13792)*
<!-- Sent from Reviewable.io -->
